database/src/backend/redb/database.rs

```diff
@@ -78,29 +187,105 @@ impl<T: Table + 'static> DatabaseRo<T> for RedbTableRw<'_, T::Key, T::Value> {
     fn get_range<'a, Range>(
         &'a self,
         range: Range,
-    ) -> Result<impl Iterator<Item = Result<T::Value, RuntimeError>> + 'a, RuntimeError>
+    ) -> Result<impl Iterator<Item = Result<(T::Key, T::Value), RuntimeError>> + 'a, RuntimeError>
     where
         Range: RangeBounds<T::Key> + 'a,
     {
         get_range::<T, Range>(self, range)
     }
+
+    #[inline]
+    fn iter(
+        &self,
+    ) -> Result<impl Iterator<Item = Result<(T::Key, T::Value), RuntimeError>> + '_, RuntimeError>
+    {
+        iter::<T>(self)
+    }
+
+    #[inline]
+    fn keys(
+        &self,
+    ) -> Result<impl Iterator<Item = Result<T::Key, RuntimeError>> + '_, RuntimeError> {
+        keys::<T>(self)
+    }
+
+    #[inline]
+    fn values(
+        &self,
+    ) -> Result<impl Iterator<Item = Result<T::Value, RuntimeError>> + '_, RuntimeError> {
+        values::<T>(self)
+    }
+
+    #[inline]
+    fn len(&self) -> Result<u64, RuntimeError> {
+        len::<T>(self)
+    }
+
+    #[inline]
+    fn first(&self) -> Result<(T::Key, T::Value), RuntimeError> {
+        first::<T>(self)
+    }
+
+    #[inline]
+    fn last(&self) -> Result<(T::Key, T::Value), RuntimeError> {
+        last::<T>(self)
+    }
+
+    #[inline]
+    fn is_empty(&self) -> Result<bool, RuntimeError> {
+        is_empty::<T>(self)
+    }
 }
 
 impl<T: Table + 'static> DatabaseRw<T> for RedbTableRw<'_, T::Key, T::Value> {
-    // `redb` returns the value after `insert()/remove()`
-    // we end with Ok(()) instead.
+    // `redb` returns the value after function calls so we end with Ok(()) instead.
 
     #[inline]
     fn put(&mut self, key: &T::Key, value: &T::Value) -> Result<(), RuntimeError> {
-        self.insert(key, value)?;
+        redb::Table::insert(self, key, value)?;
         Ok(())
     }
 
     #[inline]
     fn delete(&mut self, key: &T::Key) -> Result<(), RuntimeError> {
-        self.remove(key)?;
+        redb::Table::remove(self, key)?;
+        Ok(())
+    }
+
+    #[inline]
+    fn clear(&mut self) -> Result<(), RuntimeError> {
+        // FIXME:
+        // Deleting and re-creating the table outright
+        // would be faster, although we don't have direct
+        // access to the write transaction, which implements
+        // this functionality in `redb::WriteTransaction::delete_table`.
+
+        // Pop the last `(key, value)` until there are none.
+        while redb::Table::pop_last(self)?.is_some() {}
+
         Ok(())
     }
```

Comment:
I think we could create an intermediate `RedbTable` type like our `HeedTable` type such that it holds onto a reference of the transaction such that this would be possible.

Reply:
Yeah I think that would be better

Reply:
Actually, this is hard to impl actually since `redb` doesn't allow deleting tables that are open, [it returns `TableAlreadyOpen`](https://docs.rs/redb/2.0.0/src/redb/transactions.rs.html#380-382).

If our custom `RedbTable` itself holds onto the table we want to delete, we have to do some swapping like so:

1. swap our `self.db` table with a temporary table
2. delete the target table
3. create the target table, set `self.db` to it again
4. delete the temporary table

Only issue is that this is not `Drop`-safe, as in if we early return or panic, our `self.db` is set to the invalid temporary table. Creating a special struct and doing `fn drop() { /* reset self.db */ }` is not good either since that can error too.

I think there are only 2 safe alternatives:
1. Use `redb::Table::retain(self, false)` which hopefully compiles down efficiently since the predicate is known
2. Just remove `DatabaseRw::clear()` from this PR

Reply:
could we move this to env? then we would be able to spawn a tx to remove a table

Reply:
Sure, moved to `EnvInner::clear_db()`.